### PR TITLE
ACROSS API: TLE handler and TLE API endpoints for BurstCube and Swift

### PR DESCRIPTION
**Pull request by jak574**
## Pull Request Description

This pull request adds necessary code to fetch and store two line elements (TLEs) for missions supported by the ACROSS API. The code in this pull request also endpoints for BurstCube (with placeholder configuration using the ISS TLE) and for Swift.

## Components of this pull Request

This pull request consists of the following components:

1. Addition of a DynamoDB table called `acrossapi_tle`, which will be used to store mission TLEs for many epochs.
2. The `TLEEntry` class, a Pydantic model that also doubles as a DynamoDB model for TLEs. 
3. `TLEBase`: A base class for fetching and storing TLEs for a given mission and epoch.
4. `BurstCubeTLE` and `SwiftTLE` classes, which provide classes to fetch TLEs for these two missions. These classes inherit from `TLEBase` and otherwise only contain configuration.
5. API endpoints for querying TLE for BurstCube and Swift.

### `TLEEntry` description
The `TLEEntry` class describes a TLE with the following attributes: `satname`, `tle1`, `tle2` and `epoch`. `satname` is the standard name of the mission as given in the Satellite Catalog. `tle1` and `tle2` contains the two lines of the TLE, and `epoch` is the epoch of the TLE, derived from the TLE itself. 

`TLEEntry` provides two class methods relating to database interactions: `find_tles_between_epochs` which for a given `satname` and start and stop epochs, fetches all TLEs with epochs in that range. `write` will write the current TLEEntry to the database.

### `TLEBase` description
`TLEBase` performs the task of managing TLEs, both fetching from and if necessary, populating the TLE database. First it attempts to fetch a TLE for a given epoch from the database. If the TLE epoch is within the specified tolerance (`tle_bad`) number of days, then all good. If not, then it attempts to fetch an updated TLE from the internet, and if successful it will store that TLE in the TLE database. 

There are two types of TLE format supported:
1. 3 line Name/TLE format, as used by websites like Celestrak.
2. Concatenated TLE history files, used by HEASARC and mission operation centers, which include all TLEs from launch to now.

Through configuration you can supply URLs for these two TLE types to `TLEBase`, you can supply both, either or neither. If neither URL is provided, the code will only access the database for TLEs. This might be useful for situations where the TLE database is populated by another app. 

### API Endpoints
I have included API endpoints for two missions here to demonstrate the structure of the API code for multiple missions. With a goal of maximum code reuse, the TLE classes for each mission only consists of configuration parameters, and a caching decorator function that caches up to 128 result in memory for up to 1 hour (value picked somewhat arbitrarily), minimizing database rereads. 

For the two missions including in this pull request, other than changing Swift->BurstCube the mission specific code differs by only 5 lines, which are the configuration options for the TLE. Further missions can be easily added by cutting and pasting another mission directory, changing the mission name and updating the URLs.

## Other changes

This pull request adds the following dependencies: `architect-functions`, `boto3` (which for some reason the architect module doesn't depend on, but does require), `cachetools` which provides the time-to-live caching ability, `email-validator`, because otherwise FASTAPI complains about it not being installed constantly, and `types-cachetools` which is required for `mypy`. 

## Finally...

This description is probably longer than the pull request... happy holidays!

### CODE REVIEW

**Review by lpsinger — CHANGES_REQUESTED** [review]
High-level comments:

- Could we use astropy.time.Time rather than datetime.datetime? That will facilitate conversions between UTC, GPS, etc. and will also support vectorized time conversions in the future.
- As I understand it, https://www.space-track.org/ is a primary source and Celestrak is a secondary source. Why use Celestrak instead of Space-Track.org?
- These API endpoints should set cache-control response headers.
- There is a lot of boilerplate here as compared to the FastAPI function decorator pattern. I suppose that the boilerplate comes from Pydantic. Is there a way to cut down on the boilerplate?

**Comment by jak574** [comment]
Regarding space-track.org, the code is agnostic as to where it gets it's TLEs from, but space-track.org does require registration (so not a simple URL), and I couldn't get their API to work when I tried it.

**Comment by jak574** [comment]
Can you give me an example as to what you consider "boilerplate"?

**Comment by lpsinger** [comment]
> Regarding space-track.org, the code is agnostic as to where it gets it's TLEs from, but space-track.org does require registration (so not a simple URL), and I couldn't get their API to work when I tried it.

Registration isn't necessarily an impediment because we can save API keys in our [configuration](https://gcn.nasa.gov/docs/contributing/configuration). What issues did you have with their API?

**Comment by lpsinger** [comment]
> Can you give me an example as to what you consider "boilerplate"?

For example, we have to define a TLE subclass for each mission, _and_ write a function that instantiates it and returns a member of it.

**Comment by jak574** [comment]
> > Can you give me an example as to what you consider "boilerplate"?
> 
> For example, we have to define a TLE subclass for each mission, _and_ write a function that instantiates it and returns a member of it.

This is a design choice.

File: python/requirements.in [hunk 1]
```diff
@@ -1,4 +1,9 @@
+architect-functions
 astropy
+boto3
+cachetools
+email-validator
```
[the remark this answers is not in the document]
Comment (jak574) [reply python/requirements.in:5]: Do you mean the `email-validator`? It's unneeded for the code to run, but if you don't install it then FastAPI continously puts out this error: 

```
email-validator not installed, email fields will be treated as str.
To install, run: pip install email-validator
```
So installing it fixes this annoyance. The other two are required. FYI I don't actually import `boto3` but `architect-functions` requires it. https://github.com/architect/functions-python/pull/30

**Comment by jak574** [comment]
> * Could we use astropy.time.Time rather than datetime.datetime? That will facilitate conversions between UTC, GPS, etc. and will also support vectorized time conversions in the future.

Yep. 3664182 fixes this.

File: python/requirements.in (hunk 1 above)
[the remark this answers is not in the document]
Comment (lpsinger) [reply python/requirements.in:5]: Got it!

File: python/requirements.in [hunk 2]
```diff
@@ -1,4 +1,9 @@
...
 fastapi
 mangum
 requests
+types-cachetools
```
Comment (lpsinger) [inline python/requirements.in:9]: Technically should this go in the root requirements.txt file, as it is only a dependency for development?

File: python/lambda.py [hunk 3]
```diff
@@ -6,6 +6,9 @@
 
 from across_api.base.api import app
 from across_api.across.api import *  # noqa F401
+from across_api.burstcube.api import *  # noqa F401
+from across_api.swift.api import *  # noqa F401
```
Comment (lpsinger) [inline python/lambda.py:10]: I think that we just need to run the FastAPI function decorators; we don't actually need to pull these into the module namespace, do we?

```suggestion
import across_api.across.api  # noqa F401
import across_api.burstcube.api  # noqa F401
import across_api.swift.api  # noqa F401
```

File: python/across_api/swift/tle.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        URL to retrieve the TLE from.
+    tle_concat
+        URL to retrieve the TLE from in concatenated format.
+    tle_bad
+        If the TLE is this many days old, it is considered outdated, and a new
+        TLE will be retrieved.
+    tle_min_epoch
+        Minimum epoch for which TLEs are available, typically this will
+        correspond to a date after the launch of the spacecraft.
+    """
```
Comment (lpsinger) [inline python/across_api/swift/tle.py:33]: Do we need to copy this docstring for every mission?

Comment (jak574) [reply python/across_api/swift/tle.py]: No. Fixed in 729f3a7.

File: python/requirements.in (hunk 2 above)
Comment (jak574) [reply python/requirements.in]: Fixed in 57c8237.

**Comment by jak574** [comment]
> * These API endpoints should set cache-control response headers.

Can we defer this for now? There is a trade space here on solutions for this.

File: python/requirements.in (hunk 1 above)
[the remark this answers is not in the document]
Comment (lpsinger) [reply python/requirements.in:5]: boto3 should probably be a dev dependency; see https://github.com/architect/functions-python/pull/30#issuecomment-1872347955.

[the remark this answers is not in the document]
Comment (jak574) [reply python/requirements.in:4]: Hmm, good point. Fixed by cf28474.

**Review by lpsinger — COMMENTED** [review]
PyPI holds a variety of client libraries for [spacetrack](https://pypi.org/search/?q=spacetrack) and [celestrak](https://pypi.org/search/?q=celestrak). Did you review these? If so, what led you to write your own?

**Comment by jak574** [comment]
> PyPI holds a variety of client libraries for [spacetrack](https://pypi.org/search/?q=spacetrack) and [celestrak](https://pypi.org/search/?q=celestrak). Did you review these? If so, what led you to write your own?

Mostly just unfamiliarity and the fact that the code is pretty trivial to access those data. 

I did look into this and just pushed a commit that uses https://github.com/python-astrodynamics/spacetrack to access space-track.org data, which makes the code cleaner for sure. 

Celestrak is just downloading from a URL, so a full module for it seems like overkill.

**Review by lpsinger — APPROVED** [review]

**Comment by lpsinger** [comment]
This broke the deployment: https://github.com/nasa-gcn/gcn.nasa.gov/actions/runs/7423844220/job/20202169452

Would you please try to fix it?

**Comment by jak574** [comment]
Looking into it now.

**Comment by lpsinger** [comment]
Actually this is definitely a bug in Architect's hydration. See how it's calling `pip`?

```
pip3 install -r requirements.txt -t ./vendor --only-binary=:all: --platform=manylinux2014_aarch64 --python-version 3.11
```

**Comment by jak574** [comment]
You mean that it's doing `--only-binary=:all:`? That's certainly what's breaking it because there is no arm binary on PyPI for LogBook.

**Comment by lpsinger** [comment]
> You mean that it's doing `--only-binary=:all:`? That's certainly what's breaking it because there is no arm binary on PyPI for LogBook.

Yes. Logbook has Cython code, and it's distributed as source only. We'll need to either cross compile it or run the hydration step under qemu.

**Comment by jak574** [comment]
Seems like a solution that allows us to install source-only pip files is best in the long term, if we can do that. Short term we can either revert 9435d90 and remove this dependency. We could also ask the LogBook folks to add aarch64 builds to PyPI as they have [pretty much everything else on there](https://pypi.org/project/Logbook/#files).

**Pull request merged**